src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1748,11 +1748,21 @@ AssignTasksToConnectionsOrWorkerPool(DistributedExecution *execution)
 
 		SetAttributeInputMetadata(execution, shardCommandExecution);
 		ShardPlacement *taskPlacement = NULL;
+		int32 localGroupId = GetLocalGroupId();
 		foreach_ptr(taskPlacement, task->taskPlacementList)
 		{
 			int connectionFlags = 0;
-			char *nodeName = taskPlacement->nodeName;
-			int nodePort = taskPlacement->nodePort;
+			char *nodeName = NULL;
+			int nodePort = 0;
+			if(taskPlacement->groupId == localGroupId) {
+				nodeName = taskPlacement->nodeName;
+				nodePort = taskPlacement->nodePort;
+			}else {
+				WorkerNode *workerNode = LookupNodeForGroup(taskPlacement->groupId);
```

Comment:
This gives us a extra opportunity to get a fresh hostname from pg_dist_node. However, we may have cached pg_dist_node already, so this only works if `commit prepared` sends the correct invalidation. It's worth confirming whether this is the case.